orc/src/main/java/org/apache/iceberg/orc/OrcMetrics.java

```diff
@@ -19,51 +19,211 @@
+        Optional<ByteBuffer> orcMax = (colStat.getNumberOfValues() > 0) ?
+            fromOrcMax(icebergCol, colStat) : Optional.empty();
+        orcMax.ifPresent(byteBuffer -> upperBounds.put(icebergCol.fieldId(), byteBuffer));
+      }
+    }
+
+    return new Metrics(numOfRows,
+        columSizes,
+        valueCounts,
+        nullCounts,
+        lowerBounds,
+        upperBounds);
+  }
+
   static Metrics fromWriter(Writer writer) {
-    // TODO: implement rest of the methods for ORC metrics in
-    // https://github.com/apache/incubator-iceberg/pull/199
-    return new Metrics(writer.getNumberOfRows(),
-        null,
-        null,
-        Collections.emptyMap(),
-        null,
-        null);
+    try {
+      return buildOrcMetrics(writer.getNumberOfRows(),
+          writer.getSchema(), writer.getStatistics());
+    } catch (IOException ioe) {
+      throw new RuntimeIOException(ioe, "Failed to get statistics from writer");
+    }
+  }
+
+  private static long toMicros(Timestamp ts) {
+    return ts.getTime() * 1000;
+  }
+
+  private static Optional<ByteBuffer> fromOrcMin(Types.NestedField column,
+                                                 ColumnStatistics columnStats) {
+    ByteBuffer min = null;
+    if (columnStats instanceof IntegerColumnStatistics) {
+      IntegerColumnStatistics intColStats = (IntegerColumnStatistics) columnStats;
+      min = column.type().typeId() == Type.TypeID.INTEGER ?
+          Conversions.toByteBuffer(column.type(), (int) intColStats.getMinimum()) :
+          Conversions.toByteBuffer(column.type(), intColStats.getMinimum());
+    } else if (columnStats instanceof DoubleColumnStatistics) {
+      double minVal = ((DoubleColumnStatistics) columnStats).getMinimum();
+      min = column.type().typeId() == Type.TypeID.DOUBLE ?
+          Conversions.toByteBuffer(column.type(), minVal) :
+          Conversions.toByteBuffer(column.type(), (float) minVal);
+    } else if (columnStats instanceof StringColumnStatistics) {
+      min = Conversions.toByteBuffer(column.type(),
+          ((StringColumnStatistics) columnStats).getMinimum());
+    } else if (columnStats instanceof DecimalColumnStatistics) {
+      min = Optional
+          .ofNullable(((DecimalColumnStatistics) columnStats).getMinimum())
+          .map(minStats -> {
+            BigDecimal minValue = minStats.bigDecimalValue()
+                .setScale(((Types.DecimalType) column.type()).scale());
+            return Conversions.toByteBuffer(column.type(), minValue);
+          })
+          .orElse(null);
+    } else if (columnStats instanceof DateColumnStatistics) {
+      min = Optional.ofNullable(((DateColumnStatistics) columnStats).getMinimum())
+          .map(minStats -> Conversions.toByteBuffer(column.type(),
+              (int) ChronoUnit.DAYS.between(EPOCH_DAY,
+                  EPOCH.plus(minStats.getTime(), ChronoUnit.MILLIS).toLocalDate())))
```

Comment:
Should this conversion be in a helper method somewhere?

Reply:
Yeah, I see these constructs repeated pretty much everywhere in the codebase but I don't see helper methods. I can try to consolidate in another PR for reuse to avoid making this PR larger in scope.

Reply:
There are `DateTimeUtils` in core you could use, at least for the conversion from `LocalDate` to `int` and similar.